akka-stream-tests/src/test/scala/akka/stream/scaladsl/FoldResourceAsyncSinkSpec.scala

```diff
@@ -0,0 +1,246 @@
+        },
+        _.close()
+      )
+
+      val fut = Source.fromIterator(() ⇒ manyLines.grouped(64)).runWith(sink)
+      Await.result(fut.failed, remainingOrDefault) should be(TE(""))
+      (r.isOpened, r.isClosed) should be((true, true))
+    }
+
+    "write contents to a resource" in assertAllStagesStopped {
+      val r = new TestResource[String]()
+      val sink = Sink.foldResourceAsync[String, TestResource[String]](() ⇒ r.open(), _ write _, _.close())
+      val fut = Source.fromIterator(() ⇒ manyLines.grouped(64)).runWith(sink)
+
+      Await.result(fut, remainingOrDefault) should be(Done)
+      (r.isOpened, r.isClosed) should be((true, true))
+      r.written.mkString should be(manyLines)
+    }
+
+    "continue when Strategy is Resume and exception happened synchronously" in assertAllStagesStopped {
+      val r = new TestResource[String]()
+      val sink = Sink.foldResourceAsync[String, TestResource[String]](
+        () ⇒ r.open(),
+        (r, s) ⇒ if (s.contains("b")) throw TE("") else r.write(s),
+        _.close()
+      )
+        .withAttributes(supervisionStrategy(resumingDecider))
+      val fut = Source.fromIterator(() ⇒ manyLinesArray.iterator).runWith(sink)
+
+      Await.result(fut, remainingOrDefault) should be(Done)
+      (r.isOpened, r.isClosed) should be((true, true))
+      r.written.mkString should be(manyLinesArray.filterNot(_.contains("b")).mkString)
+    }
+
+    "continue when Strategy is Resume and exception happened asynchronously" in assertAllStagesStopped {
+      val r = new TestResource[String]()
+      val sink = Sink.foldResourceAsync[String, TestResource[String]](
+        () ⇒ r.open(),
+        (r, s) ⇒ Future(if (s.contains("b")) throw TE("")).flatMap(_ ⇒ r.write(s)),
+        _.close()
+      )
+        .withAttributes(supervisionStrategy(resumingDecider))
+      val fut = Source.fromIterator(() ⇒ manyLinesArray.iterator).runWith(sink)
+
+      Await.result(fut, remainingOrDefault) should be(Done)
+      (r.isOpened, r.isClosed) should be((true, true))
+      r.written.mkString should be(manyLinesArray.filterNot(_.contains("b")).mkString)
+    }
+
+    "continue when Strategy is Restart and exception happened synchronously" in assertAllStagesStopped {
+      val r = new TestResource[String]()
+      val sink = Sink.foldResourceAsync[String, TestResource[String]](
+        () ⇒ r.open(),
+        (r, s) ⇒ if (s.contains("b")) throw TE("") else r.write(s),
+        _.close()
+      )
+        .withAttributes(supervisionStrategy(restartingDecider))
+      val fut = Source.fromIterator(() ⇒ manyLinesArray.iterator).runWith(sink)
+
+      Await.result(fut, remainingOrDefault) should be(Done)
+      (r.isOpened, r.isClosed) should be((true, true))
+      r.written.mkString should be(manyLinesArray.filterNot(line ⇒ line.contains("a") || line.contains("b")).mkString)
+
+    }
+
+    "continue when Strategy is Restart and exception happened asynchronously" in assertAllStagesStopped {
```

Comment:
`onPush` after `onUpstreamFinish` sounds unlikely. I would guess there is a race condition in the stage somewhere with all the async callbacks going on.